### DISCUSSION TITLE
Multi Track Recording - Download ZIP file

### DISCUSSION THREAD

**Discussion by reinhardwh** [jamulussoftware/jamulus#1329] [Ideas]
Would like to record tracks individually.
Sonobus offer this feature.
Advantage: Create the mix manually

**Comment by chrisrimple** [comment]
If you set up a Jamulus Server, you can enable recording that will create WAV files for each Client connected to the Server.

If you don't own a Server, you can recording individual Clients by running multiple Clients at the same time, muting all but one musician in each Client, then directing the Jamulus output to BlackHole or another virtual audio driver and into a recording application.

**Reply by reinhardwh** [reply]
Hello Chris,

thanks for the quick reply.
Running my own server...
Can you please point me to the correct settings.

**Reply by ann0see** [reply]
Are you running it headless (CLI) or with GUI?

For headless see: https://jamulus.io/wiki/Command-Line-Options and the `-R` option

**Reply by reinhardwh** [reply]
Hello Ann0see

thanks for the tip.
Now I noticed in the description  "file(s)" - actually there are multiple files saved, for each user, thats perfect.

**Reply by ann0see** [reply]
Good to hear!

**Comment by reinhardwh** [comment]
![jamulus-remote-recording](https://user-images.githubusercontent.com/77500932/112229689-b9998980-8c33-11eb-8c56-f37eb2b0a78d.jpg)
Hello again ann0see and Chris,
used this guide to start and stop recordings: https://github.com/vdellamea/jamulus-server-remote
Works fine, without issues.
The only issue: 
* When downloading from WEB page the ZIP file cannot be extracted.
* Downloading ZIP SFTP (WINSCP) the ZIP file is ok....
What could be the reason for this?

**Reply by ann0see** [reply]
No idea. I‘d ask the person who created the web ui. Maybe it’s a Bug in his software?

**Reply by reinhardwh** [reply]
Most likely not... When downloading the ZIP file using SFTP the file can be extracted. Its just the download via WEB-page is failing...
Will contact  @vdellamea
Thanks for all the help

**Reply by ann0see** [reply]
Great!

**Comment by vdellamea** [comment]
Here I am. This issue is slightly strange and I never had it; did you try from different browsers?

**Comment by reinhardwh** [comment]
Hello @vdellamea ,
first at all: Thanks for sharing the very useful PHP scripts and the excellent documentation. Very much appreciated.

I used latest version Firefox and Chrome for downloading.

Just compared the downloaded file and the SFTP-transferred file:
Difference is directly in the beginning.... of ZIP file
![compare](https://user-images.githubusercontent.com/77500932/112323669-4aaf4580-8cb2-11eb-8477-bfe346c9d2e6.jpg)

So its not your script - but something when downloading in browser...

**Reply by vdellamea** [reply]
Actually, it is compressed using the Linux zip command, and there is nothing else done. I checked the files downloaded from this evening rehearsal: they begin with `504B0304`, like your sftp file. The added byte is a 20, that corresponds to a space; while I do not know why, this seems strange.

**Reply by hoffie** [reply]
I guess it would make sense to double check the locally used `download.php`. The version in git looks good, but maybe it contains whitespace outside of the PHP tags locally?

**Reply by reinhardwh** [reply]
> I guess it would make sense to double check the locally used `download.php`. The version in git looks good, but maybe it contains whitespace outside of the PHP tags locally?

Hello hoffie,

ZIP file on server is good.

ZIP file is corrupted when downloading via browser only.

SFTP download works fine...

**Reply by hoffie** [reply]
> ZIP file is corrupted when downloading via browser only.

Yes. That's why I was suggesting to check `download.php`. You could try running a checksum on it to verify if it is as epxected.

I think it should yield this if you are using the latest version:

```
$ sha256sum download.php 
185b2c23dfcd08423ff31c0831ce593f81e00d0d9f3914f5bf14f899617eb26d  download.php
```

**Reply by vdellamea** [reply]
@hoffie right! I forgot I was passing through the `download.php` script. Actually, it includes `config.php` that has a newline after `?>` , so this could be the issue. @reinhardwh : check at the end of `config.php`, you need to have `?>`  as the very last characters. I am modifying it (and by the way, soon I'll release the new version with automix).

**Reply by reinhardwh** [reply]
Hello @vdellamea 
screenshot of my config.php
![config-php](https://user-images.githubusercontent.com/77500932/112389093-73f0c580-8cf4-11eb-90fd-79b2ce76d7c6.jpg)

**Reply by reinhardwh** [reply]
Removed the line 12 and tested.
Browser download is corrupted
SFTP download OK
Compared the ZIP files on my local drive. Space added in file which has been downloaded via browser.

**Reply by reinhardwh** [reply]
> 
> 
> > ZIP file is corrupted when downloading via browser only.
> 
> Yes. That's why I was suggesting to check `download.php`. You could try running a checksum on it to verify if it is as epxected.
> 
> I think it should yield this if you are using the latest version:
> 
> ```
> $ sha256sum download.php 
> 185b2c23dfcd08423ff31c0831ce593f81e00d0d9f3914f5bf14f899617eb26d  download.php
> ```

Same result:
`root@rs004309:/var/www/html# sha256sum download.php
185b2c23dfcd08423ff31c0831ce593f81e00d0d9f3914f5bf14f899617eb26d  download.php
root@rs004309:/var/www/html#
`

**Reply by hoffie** [reply]
As per @vdellamea's comment, it's more likely that `config.php` is the culprit. Your screenshot does not show the problem which @vdellamea already fixed in git (leading space in the first byte), but it does show a trailing new line (and maybe space?). Try deleting the empty line 14 from `config.php`.

**Reply by reinhardwh** [reply]
Phantastic! 
Now the browser download works!
Thanks for all the help and advice.
Screenshot of corrected config.php
![config-php-fixed](https://user-images.githubusercontent.com/77500932/112391108-ac45d300-8cf7-11eb-99e9-b5aa6af8cb61.jpg)
